hugr-core/src/export.rs

```diff
@@ -767,31 +745,47 @@ impl<'a> Context<'a> {
         match t {
             SumType::Unit { size } => {
                 let items = self.bump.alloc_slice_fill_iter((0..*size).map(|_| {
-                    self.make_term(model::Term::List {
-                        items: &[],
-                        tail: None,
-                    })
+                    model::ListItem::Item(self.make_term(model::Term::List { items: &[] }))
                 }));
-                let list = model::Term::List { items, tail: None };
-                let variants = self.make_term(list);
+                let variants = self.make_term(model::Term::List { items });
                 self.make_term(model::Term::Adt { variants })
             }
             SumType::General { rows } => {
-                let items = self
-                    .bump
-                    .alloc_slice_fill_iter(rows.iter().map(|row| self.export_type_row(row)));
-                let list = model::Term::List { items, tail: None };
+                let items = self.bump.alloc_slice_fill_iter(
+                    rows.iter()
+                        .map(|row| model::ListItem::Item(self.export_type_row(row, None))),
+                );
+                let list = model::Term::List { items };
                 let variants = { self.make_term(list) };
                 self.make_term(model::Term::Adt { variants })
             }
         }
     }
 
-    pub fn export_type_row<RV: MaybeRV>(&mut self, t: &TypeRowBase<RV>) -> model::TermId {
-        let mut items = BumpVec::with_capacity_in(t.len(), self.bump);
-        items.extend(t.iter().map(|row| self.export_type(row)));
+    pub fn export_type_row<RV: MaybeRV>(
+        &mut self,
+        row: &TypeRowBase<RV>,
+        tail: Option<model::TermId>,
+    ) -> model::TermId {
+        let mut items = BumpVec::with_capacity_in(row.len() + tail.is_some() as usize, self.bump);
+
+        for t in row.iter() {
+            match t.as_type_enum() {
+                TypeEnum::RowVar(var) => {
+                    items.push(model::ListItem::Splice(self.export_row_var(var.as_rv())));
+                }
+                _ => {
+                    items.push(model::ListItem::Item(self.export_type(t)));
+                }
+            }
+        }
+
+        if let Some(tail) = tail {
```

Comment:
Consider renaming `export_type_row` to `export_type_row_with_tail` and then adding `export_type_row` that just adds the `None`